doc/doc/basics/ddlog.md

```diff
@@ -102,59 +102,69 @@ The function is supposed to take as input a triple of sentence id, array of word
+                 returns rows like people_mentions
   implementation "udf/ext_people.py" handles tsv lines.
 ```
 
 Then this user-defined function `ext_people` can be called in the following way to add more tuples to `people_mentions` taking triples from the `sentences` relation:
 
 ```
-people_mentions :- !ext_people(ext_people_input).
-ext_people_input(s, ARRAY_TO_STRING(words, "~^~"), ARRAY_TO_STRING(ner_tags, "~^~")) :-
-  sentences(_1, _2, words, _3, _4, _5, ner_tags, _6, s).
+people_mentions +=
+  ext_people(s, ARRAY_TO_STRING(words, "~^~"), ARRAY_TO_STRING(ner_tags, "~^~")) :-
+  sentences(_, _, words, _, _, _, ner_tags, _, s).
 ```
 
 In a similar way, we can have another UDF map candidate relationships and supervise them.
 
 ```
-function ext_has_spouse over like ext_has_spouse_input
-                     returns like has_spouse_candidates
+function ext_has_spouse over (sentence_id text, mid1 text, text1 text, mid2 text, text2 text)
+                     returns rows like has_spouse_candidates
   implementation "udf/ext_has_spouse.py" handles tsv lines.
 
-has_spouse_candidates :- !ext_has_spouse(ext_has_spouse_input).
-ext_has_spouse_input(s, p1_id, p1_text, p2_id, p2_text) :-
-  people_mentions(s, _1, _2, p1_text, p1_id),
-  people_mentions(s, _3, _4, p2_text, p2_id).
+has_spouse_candidates +=
+  ext_has_spouse(s, p1_id, p1_text, p2_id, p2_text) :-
+  people_mentions(s, _, _, p1_text, p1_id),
+  people_mentions(s, _, _, p2_text, p2_id).
 ```
 
 ### Feature Extraction Rules
 Also, we use a UDF to extract features for the candidate relationships.
 
 ```
-function ext_has_spouse_features over like ext_has_spouse_features_input
-                              returns like has_spouse_features
+function ext_has_spouse_features over rows like ext_has_spouse_features_input
+                              returns rows like has_spouse_features
   implementation "udf/ext_has_spouse_features.py" handles tsv lines.
 
-has_spouse_features :- !ext_has_spouse_features(ext_has_spouse_features_input).
-ext_has_spouse_features_input(ARRAY_TO_STRING(words, "~^~"), rid, p1idx, p1len, p2idx, p2len) :-
-  sentences(_1, _2, words, _3, _4, _5, _6, _7, s),
-  has_spouse_candidates(person1_id, person2_id, s, _8, rid, _9),
-  people_mentions(s, p1idx, p1len, _10, person1_id),
-  people_mentions(s, p2idx, p2len, _11, person2_id).
+function ext_has_spouse_features over (words text, rid text, position1 int, length1 int, position2 int, length2 int)
+                              returns rows like has_spouse_features
+  implementation "udf/ext_has_spouse_features.py" handles tsv lines.
+
+has_spouse_features +=
+  ext_has_spouse_features(ARRAY_TO_STRING(words, "~^~"), rid, p1idx, p1len, p2idx, p2len) :-
+  sentences(_, _, words, _, _, _, _, _, s),
+  has_spouse_candidates(person1_id, person2_id, s, _, rid, _),
+  people_mentions(s, p1idx, p1len, _, person1_id),
+  people_mentions(s, p2idx, p2len, _, person2_id).
```

Comment:
It'd be nice if we could use gist-it here as well instead of cut/pasting the code.  Keeping the line numbers in sync will be a bit tricky though..  Maybe we can put markers as comment in the actual code and grab their line numbers to always generate an accurate gist-it url to embed?